atomic_reactor/plugins/pre_reactor_config.py

```diff
@@ -37,6 +37,198 @@ def get_config(workflow):
         return conf
 
 
+def get_koji(workflow):
+    return get_config(workflow)['koji']
+
+
+def get_koji_session(workflow):
+    config = get_koji(workflow)
+
+    from atomic_reactor.koji_util import create_koji_session
+    auth_info = {
+        "proxyuser": config['auth'].get('proxyuser'),
+        "ssl_certs_dir": config['auth'].get('ssl_certs_dir'),
+        "krb_principal": config['auth'].get('krb_principal'),
+        "krb_keytab": config['auth'].get('krb_keytab_path')
+    }
+    return create_koji_session(config['hub_url'], auth_info)
+
+
+def get_pulp(workflow):
+    return get_config(workflow)['pulp']
+
+
+def get_pulp_session(workflow, logger, loglevel):
+    config = get_pulp(workflow)
+
+    from atomic_reactor.pulp_util import PulpHandler
+    return PulpHandler(workflow, config['name'], logger,
+                       pulp_secret_path=config['auth'].get('ssl_certs_dir'),
+                       username=config['auth'].get('username'),
+                       password=config['auth'].get('password'),
+                       dockpulp_loglevel=loglevel)
+
+
+def get_odcs(workflow):
+    return get_config(workflow)['odcs']
+
+
+def get_odcs_session(workflow):
+    config = get_odcs(workflow)
+
+    from atomic_reactor.odcs_util import ODCSClient
+
+    client_kwargs = {'insecure': config.get('insecure', False)}
+
+    openidc_dir = config['auth'].get('openidc_dir')
+    if openidc_dir:
+        token_path = os.path.join(openidc_dir, 'token')
+        with open(token_path, "r") as f:
+            client_kwargs['token'] = f.read().strip()
+
+    ssl_certs_dir = config['auth'].get('ssl_certs_dir')
+    if ssl_certs_dir:
+        cert_path = os.path.join(ssl_certs_dir, 'cert')
+        if os.path.exists(cert_path):
+            client_kwargs['cert'] = cert_path
+        else:
+            raise KeyError("ODCS ssl_certs_dir doesn't exist")
```

Comment:
There's no unit test coverage for this line.

Reply:
ok